### DISCUSSION TITLE
[i18n]Detecting language switch

### DISCUSSION THREAD

**Discussion by anazzani** [quasarframework/quasar#9811] [General - Components / Directives / etc]
Hi all.

I followed the instructions at https://quasar.dev/options/app-internationalization

Everything seems to be ok, except one thing.

I tried the 3 examples presented in the "How to use" section and also implemented the language switcher (it's defined in a q-toolbar inside my layout file).

When I change language, only the first two cases are automatically converted; the third example (`content: this.$t('mykey3')`) remains unchanged.

What's the recommended way to detect the switch in this case?

TIA.

Alessandro

PS: my env
```
$ quasar info

Operating System - Darwin(20.5.0) - darwin/x64
NodeJs - 14.17.0

Global packages
  NPM - 6.14.13
  yarn - 1.22.10
  @quasar/cli - 1.2.1
  @quasar/icongenie - 2.3.3
  cordova - Not installed

Important local packages
  quasar - 2.0.0 -- Build high-performance VueJS user interfaces (SPA, PWA, SSR, Mobile and Desktop) in record time
  @quasar/app - 3.0.0 -- Quasar Framework local CLI
  @quasar/extras - 1.10.7 -- Quasar Framework fonts, icons and animations
  eslint-plugin-quasar - Not installed
  vue - 3.1.1 -- The progressive JavaScript framework for buiding modern web UI.
  vue-router - 4.0.9
  vuex - Not installed
  electron - Not installed
  electron-packager - Not installed
  electron-builder - Not installed
  @babel/core - 7.14.6 -- Babel compiler core.
  webpack - 5.40.0 -- Packs CommonJs/AMD modules for the browser. Allows to split your codebase into multiple bundles, which can be loaded on demand. Support loaders to preprocess files, i.e. json, jsx, es7, css, less, ... and your custom stuff.
  webpack-dev-server - 4.0.0-beta.3 -- Serves a webpack app. Updates the browser on changes.
  workbox-webpack-plugin - Not installed
  register-service-worker - 1.7.2 -- Script for registering service worker, with hooks
  typescript - 4.2.2 -- TypeScript is a language for application scale JavaScript development
  @capacitor/core - Not installed
  @capacitor/cli - Not installed
  @capacitor/android - Not installed
  @capacitor/ios - Not installed

Quasar App Extensions
  *None installed*
```

**Comment by anazzani** [comment]
I just noticed that also happens when translating fields validation errors.

For example `:rules="[val => !!val || $t('mandatory')]"` is not changed if I switch language AFTER triggering the error.

While I understand that changing language while compiling a form is not a common scenario, I thought I'd mention it anyway in case there is something obvious I'm overlooking.

**Comment by anazzani** [comment]
The validation errors problem can be fixed using `reactive-rules`.

I'd still appreciate a suggestion about the first issue.

Thanks.

**Comment by MilosPaunovic** [comment]
I think it's a minor mistake in documentation; as I'm aware, in Vue, you don't have access to `this` reference in `data` function. So, what you could do is to define variable but assign a value in mounted (or some other lifecycle hook), like this:

```js
<script>
export default {
  data() {
    return {
      content: undefined
    }
  },
  mounted(){
    this.content = this.$t('mykey3')
  }
}
</script>
```

Or, just create a computed value:
```js
<script>
export default {
  computed: {
     content() {
       return this.$t('mykey3')
     }
  }
}
</script>
```

@rstoenescu @hawkeye64 If any you guys could confirm, I'd be happy to submit a PR for the docs fix.

**Reply by anazzani** [reply]
Thanks for your reply.

Unfortunately it doesn't help: the variables are correctly showing with the default language when loading the component.

The problem is that when I change language they don't update, even if I initialise them in the created or mounted lifecycle.

**Reply by MilosPaunovic** [reply]
Have you made sure that you have proper key in all of the translation files? That might cause issues.

**Reply by anazzani** [reply]
Translation files are ok.

For example I'm using a `q-option-group` for a couple of radio buttons:

```
<q-option-group
  :options="myoptions"
  type="radio"
  v-model="mymodel" />

<pre>
check: {{$t('option1')}}
</pre>
.
.
.
  mounted () {
    this.myoptions = [
      { label: this.$t('option1'), value: 1 },
      { label: this.$t('option2'), value: 2 }
    ]
```

When I switch language the radio button options do NOT change, the test inside `<pre>` does.

**Reply by hawkeye64** [reply]
> If any you guys could confirm, I'd be happy to submit a PR for the docs fix.

Sorry, I don't know much about i18n. I have never personally needed it yet.

**Reply by pdanpdan** [reply]
You put the final translated text in myoptions. What would you expect to happen when you change language?

Make it a computed and it will work.

**Reply by anazzani** [reply]
I feel stupid. :(

Apologies and thanks for your suggestion @pdanpdan, working now.

**Reply by sandimangiwa** [reply]
> You put the final translated text in myoptions. What would you expect to happen when you change language?
> 
> Make it a computed and it will work.
if I want to override the i18n property on a method with the condition this.content= this.$('mykey4') and this.content is being computed, how to do it?
```
<script>
export default {
  computed: {
     content() {
       return this.$t('mykey3')
     }
  },
  methods:{
      changeContent () {
        this.content = this.$t('mykey4')
      },
  }
}
</script>
```
